File: Data_Structures_and_Libraries/segment_tree.py
class segment_tree:
	def __init__(self, A):
		self.A = A
		self.st = [0] * 4 * len(A)
		self.build(0, 0, len(A)-1, A)


	def build(self, u, l, r, A):
		print(u, l, r)
		if l == r:
			self.st[u] = A[l]
			return

		m = int((l + r) / 2)

		self.build(2*u+1, l, m, A)
		self.build(2*u+2, m+1, r, A)

		self.st[u] = self.st[2*u+1] + self.st[2*u+2]


	def update(self, i, v):
		self.update_recursion(0, 0, len(self.A) - 1, i, v)

	def update_recursion(self, u, l, r, i, v):
		if i < l or r < i: return

		if l == r:
			self.st[u] = v
			return

		m = int((l + r) / 2)

		self.update_recursion(2*u+1, l, m, i, v)
		self.update_recursion(2*u+2, m+1, r, i, v)

		self.st[u] = self.st[2*u+1] + self.st[2*u+2]


	def querry(self, a, b):
		return self.querry_recursion(0, 0, len(self.A)-1, a, b)

	def querry_recursion(self, u, l, r, a, b):
		if b < l or r < a: return 0

		if a <= l and r <= b: return self.st[u]

		m = int((l + r) / 2)

		if b < m:
			return self.querry_recursion(2*u+1, l, m, a, b)

		elif m < a:
			return self.querry_recursion(2*u+2, m+1, r, a, b)

		else:
			left = self.querry_recursion(2*u+1, l, m, a, b)
			right = self.querry_recursion(2*u+2, m+1, r, a, b)
			return left + right


if __name__ == "__main__":
	A = [1, 2, 2, 3, 5, 4, 1, 0, 4]

	st = segment_tree(A)
	print(st.st)

	st.update(0, 5)
	print(st.st)

	print(st.querry(0, 0))
	print(st.querry(2, 5))
	print(st.querry(0, 8))
	print(st.querry(6, 8))

File: Data_Structures_and_Libraries/segment_tree.cpp
#include <bits/stdc++.h>
using namespace std;

const int NEUTRAL = 0;

int comp(int a, int b){
	return a + b;
}

class SegmentTree{
	private:
		vector<int> tree;
		int n;

		void build(int p, int l, int r, vector<int>& arr){
			if(l == r){
				tree[p] = arr[l];
				return;
			}
			int m = (l+r)/2;
			build(2*p, l, m, arr);
			build(2*p+1, m+1, r, arr);
			tree[p] = comp(tree[2*p], tree[2*p+1]);
		}

		void update(int p, int l, int r, int i, int k){
			if(i > r || i < l) return;
			if(l == r){
				//tree[p] += k;
				tree[p] = k;
				return;
			}
			int m = (l+r)/2;
			update(2*p, l, m, i, k);
			update(2*p+1, m+1, r, i, k);
			tree[p] = comp(tree[2*p], tree[2*p+1]);
		}

		int query(int p, int l, int r, int a, int b){
			if(b < l || a > r) return NEUTRAL;
			if(a <= l && b >= r) return tree[p];
			int m = (l+r)/2;
			return comp(query(2*p, l, m, a, b), query(2*p+1, m+1, r, a, b));
		}

	public:
		SegmentTree(vector<int> arr){
			n = arr.size();
			tree.assign(4*n, NEUTRAL);
			build(1, 0, n-1, arr);
		}

		int query(int a, int b){
			return query(1, 0, n-1, a, b);
		}

		void update(int i, int k){
			update(1, 0, n-1, i, k);
		}
};


int main(){
	vector<int> test;
	test.push_back(0);
	test.push_back(3);
	test.push_back(4);
	test.push_back(2);
	test.push_back(1);
	test.push_back(9);
	test.push_back(99);
	test.push_back(5);
	test.push_back(4);
	test.push_back(3);
	test.push_back(1);
	test.push_back(0);
	SegmentTree st(test);
	cout << st.query(1, 4) << endl;
	cout << st.query(3, 7) << endl;
	st.update(6, 1);
	cout << st.query(3, 7) << endl;
	return 0;
}
